### ISSUE TITLE
Changing remote debugging deprecation

### ISSUE THREAD

**Issue by ochameau** [devtools-html/rfcs#51]
**tl;dr; Deprecate connecting to remote devices (Firefox for Android, Gecko View, ...) using versions 3 or 4 versions older. (ex: Firefox 63 [nightly] would reject connecting to Firefox for Android to 61[release] or 60 but not 59 and older)** 

A warning is currently displayed in WebIDE when connecting to Firefox for Android or any other remote runtime (geckoview, firefox-os, ...) that has a deprecated version:

> The connected runtime has an old version XXX. The minimum supported version is YYY. This is an unsupported setup and may cause DevTools to fail. Please update the connected runtime.

It was originally introduced in [bug 1405008.](https://bugzilla.mozilla.org/show_bug.cgi?id=1405008)
Back then, we arbitrarily said that DevTools would officially support versions up to current ESR.
Today, ESR is 52, but soon to become 60. ([bug 1487337](https://bugzilla.mozilla.org/show_bug.cgi?id=1487337) is opened to bump the warning to 60)

But the original choice of supporting backward compatibility up to current ESR hasn't been really discussed, nor fully thought through.
ESR is something specific to Firefox for desktop. AFAIK, we don't maintain/ship any ESR on mobile. We don't have any ESR for Firefox for Android.
And given that this backward compatibility rule is about the remote runtime, it is should be defined around the mobile products rather than desktop.
So ESR doesn't really make sense regarding backward compatibility. It somewhat matches our needs because it is about "some previous release". 
But ESR lifespan is very long. It last for 8 regular firefox release. It means that we have to support up to 8 releases in the past. That's a lot! And if you wait for the additional two "qualify" releases it is up to 10 releases in the past to support!!
See this [diagram](https://www.mozilla.org/media/img/firefox/organizations/release-overview-high-res.d3c9788f2dd1.png).

This explains why today, Firefox 63 has to support up to Firefox 52!

I recently went through telemetry's data about remote debugging. See [this spreadsheet](https://docs.google.com/spreadsheets/d/19u57uS5PcnTleCwaLp3va0YDkn65LFoQmORB2uDoBxc/edit?usp=drivesdk) for more details.
And the conclusion is that the majority of our users either debug the current release version (from dev-edition) or the same version (from Nightly).
If means that we could limit our backward compatibility up to the current release version, or one version older than the current release, to accommodate people not updating their mobile firefox immediately. 
We would switch from supporting up to 10 releases in the past down to only 3 or 4.

The backward compatibility would better match remote debugging usage and be bumped on every release instead of every 10 releases. 

Reducing the compatibility range would help removing more workarounds, more frequently. We almost never do that.

It may also better reflect the reality of our compatibility. We have no tests ensuring that tools are working against any previous version, even just the previous one. Nor do we have clear review guidelines to ensure that modifications shouldn't break the support of old runtimes.

Also, with fission coming, we are going to seriously complexity some client codepaths. Having various compatibility wrappers in the way would unnecessarily complexity our codebase.
This new restriction may help but also might not be enough. I'd like to suggest a one time drop in backward compatibility,  but let's not conflate the too topics.

Finally, I can't talk about improving backward compatibility without mentioning getting rid of it by using a client synced with the server.
Either by hosting many versions of the DevTools frontend, or serving it from the remote device.
I consider that we are far from being able to do that. That would be another multi-quarter, multi-persons goals. It will be easier to start thinking about that once we got rid of XUL from DevTools, which is planned by the end of 2018.
Here, I want to suggest a pragmatic, immediately actionable change that doesn't engage any resource.

++
Alex

**Comment by juliandescottes** [comment]
I agree. I think supporting the 3 previous versions would be enough to cover the whole range of Firefox versions that are available in the mainstream at the same time.

Since we don't have tests, we should start showing the warning message as soon as the user connects to any different version, even if our policy is to "try" to support 3 versions prior. That's probably more a topic for the new remote-debugging solution.

**Comment by jryans** [comment]
I would definitely recommend reducing the compat window... As your data shows, it's quite rarely used. Certainly as a start, drop the ESR support. Maybe in the future try something more extreme. The time is wasted on compat wrappers is not worth it when there is so much else to do...

**Comment by jimblandy** [comment]
It sounds good to me. We're almost certainly not doing a good job supporting older versions anyway, given that we don't do any inter-version testing.

**Comment by digitarald** [comment]
Good from the product side. If users want to go further back, we can recommend them to use ESR.

**Comment by ochameau** [comment]
Thanks for your feedback!

So, I've updated [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1487337) to bump our compatibility threshold up to 61 or 62. Feel free to comment here or there about the best option.

To followup on Julian's suggestion, I've opened [this other bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1492187) in order to display a warning when connecting to a "compatible" version that isn't the same version as the client. That, to reflect our lack of tests.

Last, but not least, I was wondering if, instead of displaying a warning message when connecting to an unsupported version, we should prevent the connection?

**Comment by ochameau** [comment]
I opened [a third bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1492464) in order to reject the connection when connecting to an old/unsupported runtime instead of only printing a warning message and let open a half-working half-broken toolbox.

**Comment by juliandescottes** [comment]
This was effectively accepted, bugs were created by @ochameau and are tracked in the remote debugging backlog. Closing.

**Comment by captainbrosset** [comment]
I was trying to remember what our current backward compat policy was and couldn't, so I was happy when I found this RFC. Unfortunately, I feel like it fails at making it clear what the policy is today and I think it should.
@ochameau proposed going back 3 or 4 versions:

> We would switch from supporting up to 10 releases in the past down to only 3 or 4.

Then people agreed on 3, it seems:

> I agree. I think supporting the 3 previous versions would be enough to cover the whole range of Firefox versions that are available in the mainstream at the same time.

The [bug filed and fixed by Alex](https://bugzilla.mozilla.org/show_bug.cgi?id=1487337) does show that we support `N → N`, `N → N-1` and `N → N-2` (specifically [this line of code](https://searchfox.org/mozilla-central/rev/330daedbeac2bba296d663668e0e0cf248bc6823/devtools/shared/client/debugger-client.js#39)).

Is this correct?
Do we need to update our [backward compatibility documentation](https://docs.firefox-dev.tools/backend/backward-compatibility.html) to reflect this?

**Comment by juliandescottes** [comment]
Yes this is correct, we support N, N-1, N-2. I think this is aligned with the page you linked:

"Nightly desktop client MUST maintain existing compatibility back to release channel servers."
https://searchfox.org/mozilla-central/source/devtools/docs/backend/backward-compatibility.md#13

Maybe we can rephrase it to be clearer, and also make it even clearer in the `Removing old backward compatibility code` section. That should be the main thing of this section "Any backward compatibility code for version Firefox N-3 or older can and should be removed." or something like that.
